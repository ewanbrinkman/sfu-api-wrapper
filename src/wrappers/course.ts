import { requestSFUAcademicCalendarApiCourses } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawApiCourse,
} from '@api-types';
import { Course } from '@api';

export default async function course(
    department: string,
    number: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<Course> {
    const rawApiCourse: RawApiCourse = (
        await requestSFUAcademicCalendarApiCourses(
            year,
            term,
            department,
            number,
        )
    ).data;

    return Course.fromRawApiCourse(rawApiCourse);
}
