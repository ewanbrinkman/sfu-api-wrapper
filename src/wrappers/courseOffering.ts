import { requestSFUCourseOutlinesApi } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawApiCourseOffering,
} from '@api-types';
import { CourseOffering } from '@api';

export default async function courseOutline(
    department: string,
    number: number,
    section: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<CourseOffering> {
    const rawApiCourseOffering: RawApiCourseOffering = (
        await requestSFUCourseOutlinesApi(
            year,
            term,
            department,
            number,
            section,
        )
    ).data;

    return CourseOffering.fromRawApiCourseOffering(rawApiCourseOffering);
}
