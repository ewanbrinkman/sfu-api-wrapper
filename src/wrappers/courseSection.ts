import { requestSFUCourseOutlinesApi } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawCourseSectionData,
} from '@api-types';
import { CourseSection } from '@api';

export default async function courseSection(
    department: string,
    number: string,
    section: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<CourseSection> {
    const response = await requestSFUCourseOutlinesApi(
        year,
        term.toLowerCase(),
        department.toLowerCase(),
        number.toLowerCase(),
        section.toLowerCase(),
    );
    const rawCourseSectionData: RawCourseSectionData = await response.json();

    return CourseSection.fromRawCourseSectionData(
        rawCourseSectionData,
        year,
        term,
        department,
    );
}
