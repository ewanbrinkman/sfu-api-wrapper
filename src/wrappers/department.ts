import { requestSFUAcademicCalendarApiCourses, limitConcurrency } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawDepartmentCourse,
} from '@api-types';

export default async function department(
    department: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<string[]> {
    const rawDepartmentCourses: RawDepartmentCourse[] = (
        await requestSFUAcademicCalendarApiCourses(year, term, department)
    ).data;

    return rawDepartmentCourses.map(rawDepartmentCourse => rawDepartmentCourse.value);
}
