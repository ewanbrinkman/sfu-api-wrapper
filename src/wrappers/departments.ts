import { requestSFUAcademicCalendarApiCourses } from '@utils';
import { CourseOutlinesTerm, CourseOutlinesYear } from '@api-types';

type DepartmentValue = {
    title: string;
    value: string;
};

type DepartmentValues = DepartmentValue[];

export default async function departments(
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<string[]> {
    const response = (
        await requestSFUAcademicCalendarApiCourses(year, term)
    );
    const rawDepartments: DepartmentValues = await response.json();

    return rawDepartments.map(rawDepartment => rawDepartment.value);
}
