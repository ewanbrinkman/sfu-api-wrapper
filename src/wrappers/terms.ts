import { requestSFUCourseOutlinesApi } from '@utils';
import { CourseOutlinesYear, RawTerm, Term } from '@api-types';

export default async function terms(
    year: CourseOutlinesYear = 'current',
): Promise<Term[]> {
    const response = await requestSFUCourseOutlinesApi(year);
    const rawTerms: RawTerm[] = await response.json();

    return rawTerms.map((rawTerm) => rawTerm.value);
}
