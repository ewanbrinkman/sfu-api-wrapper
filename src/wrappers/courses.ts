import { requestSFUAcademicCalendarApiCourses, limitConcurrency } from '@utils';
import type {
    CourseOutlinesYear,
    CourseOutlinesTerm,
    RawDepartmentCourse,
} from '@api-types';
import { Course } from '@api';
import wrappers from '@wrappers';

export default async function courses(
    department: string,
    year: CourseOutlinesYear = 'current',
    term: CourseOutlinesTerm = 'current',
): Promise<Course[]> {
    const response = (
        await requestSFUAcademicCalendarApiCourses(year, term, department)
    );
    const rawDepartmentCourses: RawDepartmentCourse[] = await response.json();

    // Can't request all of them at once, since that spams the API and the API
    // has problems with that.
    return await limitConcurrency(
        rawDepartmentCourses,
        async (rawDepartmentCourse) => {
            return wrappers.course(department, rawDepartmentCourse.value, year, term);
        }
    );
}
