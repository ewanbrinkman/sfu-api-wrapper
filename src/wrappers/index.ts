import course from './course';
import courseSection from './courseSection';
import departmentCourseNumbers from './departmentCourseNumbers';
import departmentCourses from './departmentCourses';
import departmentNames from './departments';
import departments from './departments';
import years from './years';
import terms from './terms';

export default {
    course: course,
    courseSection: courseSection,
    departmentCourseNumbers: departmentCourseNumbers,
    departmentCourses: departmentCourses,
    departmentNames: departmentNames,
    departments: departments,
    terms: terms,
    years: years,
};
