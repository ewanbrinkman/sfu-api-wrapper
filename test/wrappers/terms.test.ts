import wrappers from '../../src/wrappers';

describe('terms', () => {
    test('request terms', async () => {
        const terms = await wrappers.terms(2021);
        expect(terms).toMatchSnapshot();
    }, 30000);
});
