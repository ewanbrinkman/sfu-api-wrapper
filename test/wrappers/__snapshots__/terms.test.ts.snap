// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terms request terms 1`] = `
[
  "fall",
  "spring",
  "summer",
]
`;
